lib/rules/define-props-declaration.js

```diff
@@ -27,20 +198,87 @@ module.exports = {
   },
   /** @param {RuleContext} context */
   create(context) {
+    const sourceCode = context.getSourceCode()
+
     const scriptSetup = utils.getScriptSetupElement(context)
     if (!scriptSetup || !utils.hasAttribute(scriptSetup, 'lang', 'ts')) {
       return {}
     }
 
     const defineType = context.options[0] || 'type-based'
+    const separateInterface = context.options[1]?.separateInterface || false
+
     return utils.defineScriptSetupVisitor(context, {
-      onDefinePropsEnter(node) {
+      onDefinePropsEnter(node, props) {
         switch (defineType) {
           case 'type-based': {
             if (node.arguments.length > 0) {
               context.report({
                 node,
-                messageId: 'hasArg'
+                messageId: 'hasArg',
+                *fix(fixer) {
+                  try {
+                    const propTypes = props.map((prop) =>
+                      getComponentPropData(prop, sourceCode)
+                    )
+
+                    const definePropsType = `{ ${propTypes
+                      .map(
+                        ({ name, type, required, defaultValue }) =>
+                          `${name}${
+                            required === false || defaultValue ? '?' : ''
+                          }: ${type}`
+                      )
+                      .join(', ')} }`
```

Comment:
Can you extract some variables here to make the template strings easier to read?

Reply:
You refactored some code here, but it is still quite unreadable because the template expressions span multiple lines. I think something like this is much more readable:

```js
const definePropTypes = propTypes.map(
  ({ name, type, required, defaultValue }) => {
    const optionalSuffix = required === false || defaultValue ? '?' : ''
    return `${name}${optionalSuffix}: ${type}`
  }
)
const definePropTypesString = `{ ${definePropTypes.join(', ')} }`
```

Reply:
@FloEdelmann How about now?